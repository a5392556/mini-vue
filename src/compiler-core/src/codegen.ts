import { NodeTypes } from "./ast";
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelper";

export function generate(ast) {
    const context = createCodegenContext();
    const { push } = context;
    genFunctionPreamble(ast, context);
    const functionName = "render";
    const args = ["_ctx", "_cache"];
    const signature = args.join(',');
    push(`function ${functionName}(${signature}) {`);
    push('return ');
    // debugger
    // console.log(ast.codegenNode);
    genNode(ast.codegenNode, context);
    push('}');
    return {
        code: context.code
    }
}

function genFunctionPreamble(ast, context) {
    const aliasHelpers = s => `${helperMapName[s]}: _${helperMapName[s]}`;
    const VueBinging = 'Vue';
    if (ast.helpers.length > 0) context.push(`const { ${ast.helpers.map(aliasHelpers).join(', ')} } = ${VueBinging}`);
    context.push('\n');
    context.push("return ");
}

function genNode(codegenNode: any, context) {
    if (!codegenNode) return;

    switch (codegenNode.type) {
        case NodeTypes.TEXT:
            genText(context, codegenNode);
            break;
        case NodeTypes.INTERPOLATION:
            genInterpolation(context, codegenNode);
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(context, codegenNode);
            break;
        case NodeTypes.ELEMENT:
            genElement(context, codegenNode);
            break;
        case NodeTypes.COMPOUND_EXPRESSION:
            genCompoundExpression(context, codegenNode);
            break;
        default:
            break;
    }
}

function isString(node) {
    return typeof node === 'string';
}

function genCompoundExpression(context: any, node: any) {
    const { children } = node;
    const { push } = context;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isString(child)) {
            push(child);
        } else {
            genNode(child, context);
        }
    }
}
function genElement(context, codegenNode) {
    const { push, helper } = context;
    const { tag, children, props } = codegenNode;
    push(`${helper(CREATE_ELEMENT_VNODE)}(`);
    // genNode(children, context);
    genNodeList(genNullable([tag, props, children]), context);
    push(')');
}

function genNodeList(nodes: any[], context) {
    const {push} = context;
    for(let i = 0; i < nodes.length; i++) {
        const node = nodes[i];
        if(isString(node)) {
            push(node);
        }else {
            genNode(node, context);
        }
        if(i < nodes.length - 1) push(', ');
    }
}
function genText(context: any, codegenNode: any) {
    const { push } = context;
    push(`'${codegenNode.content}'`);

}
function genInterpolation(context: any, codegenNode: any) {
    const { push, helper } = context;
    push(`${helper(TO_DISPLAY_STRING)}(`);
    genNode(codegenNode.content, context);
    push(')');
}
function genExpression(context: any, codegenNode: any) {
    const { push } = context;
    push(`${codegenNode.content}`);
}
function createCodegenContext() {
    const context = {
        code: '',
        push(source: string) {
            context.code += source;
        },
        helper(key) {
            return `_${helperMapName[key]}`
        }
    };
    return context;
}






function genNullable(args: any[]) {
    // throw new Error("Function not implemented.");
    return args.map(arg => arg || 'null');
}

