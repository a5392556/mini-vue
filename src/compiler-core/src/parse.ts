import { NodeTypes } from "./ast";

const enum TagType {
    START,
    END
}

export function baseParse(content: string) {
    const context = createParserContext(content);
    return createRoot(parseChildren(context, []));
}

function parseChildren(context, ancestors) {
    const nodes: any = [];
    while (!isEnd(context, ancestors)) {
        let node;
        const s = context.source;
        if (s.startsWith('{{')) {
            node = parseInterpolation(context);
        } else if (s[0] === '<') {
            if (/([a-z]*)/i.test(s[1]))
                node = parseElement(context, ancestors);
        }
        if (!node) node = parseText(context);
        nodes.push(node);
    }
    return nodes;
}


function parseElement(context: any, ancestor): any {
    const element: any = parseTag(context, TagType.START);
    ancestor.push(element);
    element.children = parseChildren(context, ancestor);
    ancestor.pop();
    // console.log(element.tag);
    // console.log(context.source.slice(2, 2 + context.source.length));
    if (startsWidthEndTagOpen(context.source, element.tag)) {
        parseTag(context, TagType.END);
    } else {
        throw new Error(`缺少结束标签:${element.tag}`);
    }
    return element;
}

function parseTag(context: any, type: TagType) {

    const match: any = /^<\/?([a-z]*)/i.exec(context.source);
    const tag = match[1];

    advanceBy(context, match[0].length);
    advanceBy(context, 1);

    return type === TagType.START
        ? {
            type: NodeTypes.ELEMENT,
            tag
        }
        : undefined;
}

function parseInterpolation(context) {

    const openDelimiter = '{{';
    const closeDelimiter = "}}";
    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length);
    advanceBy(context, openDelimiter.length);

    const rawContentLength = closeIndex - openDelimiter.length;
    const rawContent = parseTextData(context, rawContentLength);
    const content = rawContent.trim();
    advanceBy(context, closeDelimiter.length);
    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: content
        }
    }
}

function advanceBy(context: any, index: number) {
    context.source = (context.source as string).slice(index);
}

function createRoot(children) {
    return {
        children
    };
}

function createParserContext(content: string) {
    return {
        source: content
    };
}


function parseText(context: any): any {
    let endIndex = context.source.length;
    let endToken = ['<', '{{'];
    for (let i = 0; i < endToken.length; i++) {
        const index = context.source.indexOf(endToken[i]);
        if (index !== -1 && endIndex > index) {
            endIndex = index;
        }
    }

    const content = parseTextData(context, endIndex);
    return {
        type: NodeTypes.TEXT,
        content
    };
}

function parseTextData(context, length: number) {
    const content = context.source.slice(0, length);
    advanceBy(context, length);
    return content;
}

function isEnd(context: any, ancestors: any[]) {
    const s = context.source;
    if (!s) return true;

    if (s.startsWith('</')) {
        for (let i = ancestors.length -1; i >= 0; i--) {
            const tag = ancestors[i].tag;
            if (startsWidthEndTagOpen(s, tag)) {
                return true;
            }
        }
    }

    return !s;
}

function startsWidthEndTagOpen(s: string, tag: string) {
    return s.startsWith('</') && s.slice(2, tag.length + 2).toLowerCase() === tag.toLowerCase();
}

