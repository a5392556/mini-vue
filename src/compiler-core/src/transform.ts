import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelper";

export function transform(root, options?: any) {
    const context = createTransformContext(root, options);
    traverseNode(root, context);
    createRootCodegen(root);
    root.helpers = [...context.helpers.keys()];
}

function traverseNode(node, context) {
    const type: NodeTypes = node.type;

    // 遍历调用所有的 nodeTransforms
    // 把 node 给到 transform
    // 用户可以对 node 做处理
    const nodeTransforms = context.nodeTransforms;
    const exitFns: any = [];
    for (let i = 0; i < nodeTransforms.length; i++) {
      const transform = nodeTransforms[i];
  
      const onExit = transform(node, context);
      if (onExit) {
        exitFns.push(onExit);
      }
    }
  
    switch (type) {
      case NodeTypes.INTERPOLATION:
        // 插值的点，在于后续生成 render 代码的时候是获取变量的值
        context.helper(TO_DISPLAY_STRING);
        break;
  
      case NodeTypes.ROOT:
      case NodeTypes.ELEMENT:
  
        traverseChildren(node, context);
        break;
  
      default:
        break;
    }
  
  
  
    let i = exitFns.length;
    // i-- 这个很巧妙
    // 使用 while 是要比 for 快 (可以使用 https://jsbench.me/ 来测试一下)
    while (i--) {
      exitFns[i]();
    }
}

function createTransformContext(root, options) {

    const context = {
        root,
        nodeTransforms: options?.nodeTransforms || [],
        helpers: new Map(),
        helper(key) {
            context.helpers.set(key, 1);
        }
    };
    return context;
}

function traverseChildren(node, context) {
    const children = node.children;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        traverseNode(child, context);
    }
}

function createRootCodegen(root: any) {
    const { children } = root;

    // 只支持有一个根节点
    // 并且还是一个 single text node
    const child = children[0];

    if (child.type === NodeTypes.ELEMENT && child.codegenNode) {
        const codegenNode = child.codegenNode;
        root.codegenNode = codegenNode;
    } else {
        root.codegenNode = child;
    }

    // root.codegenNode = child;
    
}
