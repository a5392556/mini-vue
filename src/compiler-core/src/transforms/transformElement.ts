import { createVNodeCall, NodeTypes } from "../ast";

export function transformElement(node, context) {

    if (node.type === NodeTypes.ELEMENT) {
        return () => {
            // 中间处理层

            // tag
            const vnodeTag = `'${node.tag}'`;
            // props
            let vnodeProps;
            // chldren
            let vnodeChildren = node.children[0];
            node.codegenNode = createVNodeCall(context, vnodeTag, vnodeProps, vnodeChildren);
        }


    }
}