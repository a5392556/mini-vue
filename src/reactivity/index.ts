export function add(a: number, b: number) {
    return a + b;
}

export function handleChange(oldValue, newValue) {
    return !Object.is(oldValue, newValue);
}
