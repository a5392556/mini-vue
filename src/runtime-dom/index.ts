import { createRenderer } from '../runtime-core';
const isOn = name => /^on[A-Z]/.test(name);
function createElement(type) {
    return document.createElement(type);
}
function patchProp(el, key,oldVal ,newVal) {
    if (isOn(key)) {
        const event = key.slice(2).toLowerCase();
        el.addEventListener(event, newVal);
    } else {
        if(!newVal) el.removeAttribute(key);
        else el.setAttribute(key, newVal);
    }
}
function insert(el, container) {
    container.append(el);
}

function setElementText(el, text) {
    el.textContent = text;
}

function remove(el) {
    const parent = el.parentNode;
    parent && parent.removeChild(el);
}

const renderer: any = createRenderer({
    createElement,
    patchProp,
    insert,
    remove,
    setElementText
});

export function createApp(...args) {
    return renderer.createApp(...args);
}

export * from '../runtime-core';