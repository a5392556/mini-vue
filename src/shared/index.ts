export const extend = Object.assign;

export const isObject = val => {
    return val !== null && typeof val === 'object';
}

export const hasOwn = (target, key) => Object.prototype.hasOwnProperty.call(target, key);