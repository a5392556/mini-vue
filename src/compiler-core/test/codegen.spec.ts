import { generate } from "../src/codegen";
import { baseParse } from "../src/parse"
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";

describe('code genertor', () => {
    it('string', () => {
        const ast = baseParse('hi');
        transform(ast);
        const {code} = generate(ast);
        expect(code).toMatchSnapshot();
    });
});

describe('code genertor', () => {
    it('interpolation', () => {
        const ast = baseParse('{{message}}');
        transform(ast, {
            nodeTransform: [transformExpression]
        });
        const {code} = generate(ast);
        expect(code).toMatchSnapshot();
    });
});