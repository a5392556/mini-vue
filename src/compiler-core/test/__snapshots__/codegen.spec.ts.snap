// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code genertor element 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx,_cache) {return _createElementVNode('div', null, null)}"
`;

exports[`code genertor element and interpolation 1`] = `
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx,_cache) {return _createElementVNode('div', null, 'hi,'+_toDisplayString(_ctx.msg))}"
`;

exports[`code genertor interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx,_cache) {return _toDisplayString(message)}"
`;

exports[`code genertor string 1`] = `
"
return function render(_ctx,_cache) {return 'hi'}"
`;
